syntax = "proto3";

// package database;

service Database {
    rpc shutdown(Empty) returns (Success) {}

    rpc ping(Empty) returns (Success) {}

    rpc read(Location) returns (Cell) {}

    rpc getVersionNumber(Location) returns (Cell) {}

    rpc write(WriteRequest) returns (Success) {}

    rpc cput(CPutRequest) returns (Success) {}

    rpc erase(Location) returns (Success) {}

    rpc getRows(RowRange) returns (Tablet) {}
}

message Empty {}

message Success {
    bool success = 1;
}

message Location {
    string row = 1;
    string col = 2;
}

message Cell {
    bytes contents = 1;
    uint64 version = 2;
}

message WriteRequest {
    Location loc = 1;
    Cell cell = 2;
}

message CPutRequest {
    Location loc = 1;
    Cell curr = 2;
    Cell new_val = 3;
}

message RowRange {
    // Note that these indices are 0-indexed
    uint64 start_row = 1;
    uint64 end_row = 2;
}

message Row {
    map<string, Cell> row = 1;
}

message Tablet {
    map<string, Row> tablet = 1;
}